osu.Game.Rulesets.Osu/OsuInputManager.cs

```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
             return base.Handle(e);
         }
 
+        private OsuAction getActionForTouchSource(TouchSource source)
+        {
+            int sourceNumber = numberedTouchSources[source];
+            return sourceNumber % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
+        }
+
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (!AllowUserCursorMovement)
+            var source = e.Touch.Source;
+
+            int touchNumber = numberedTouchSources[source];
+
+            if (touchNumber > simultaneous_touches_limit)
+                return false;
+
+            var activeSources = CurrentState.Touch.ActiveSources;
+
+            var cursorTouch = activeSources.Any() ? activeSources.First() : source;
+
+            var activeTapTouches = AllowUserCursorMovement ? activeSources.Skip(1) : activeSources;
+            var limitedActiveTapTouches = activeTapTouches.Take(tap_touches_limit);
+
+            var newActiveTapActions = limitedActiveTapTouches.Select(getActionForTouchSource);
+            var newInactiveTapActions = PressedActions.Where(a => !newActiveTapActions.Contains(a)).ToList();
```

Comment:
As mentioned above, this should be saved in private fields rather than reallocating every time a touch changes state.